### ISSUE TITLE
nix-build: "invalid ownership on file ‘...perl-5.22.2/bin/perlbug’"

### ISSUE THREAD

**Issue by ilyaigpetrov** [NixOS/nix#1243]
Here are steps to reproduce it:

1. `git clone --depth 1 -b release-16.09 --single-branch https://github.com/NixOS/nixpkgs.git`
2. Boot from NixOS live iso, I used `nixos-minimal-16.09.1694.019dcc3-i686-linux.iso`, but you may use graphical x64 (tested it too), because copypasting is easier inside KDE konsole.
3. I booted in BIOS mode (not UEFI).
4. Start building:
```sh-session
cd nixpkgs
export NIXPKGS=`pwd`
export NIX_PATH=$NIXPKGS
nix-build -A perl '<default.nix>'
```
5. After long time building fails with error:
```sh-session
# 1. Copypasted from 16.09 (tag), not release-16.09, but I checked both -- both fail.
# 2. Output is from building whole NixOS ISO system, not perl alone.
gzipping man pages in /nix/store/s0wng1rn5pnddrigjjqszxc5bip4j46i-perl-5.22.2-devdoc
invalid ownership on file ‘/nix/store/qj1p9cdc9h4fhcz75qvk49azzq3gx4y0-perl-5.22.2/bin/perlbug’
cannot build derivation ‘/nix/store/fa3aq45swqqfy8zi3gmpxnzdzji7w0hl-bison-3.0.4.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/ddsgnsq8d88mimgd70fwf4jl2irhnd73-coreutils-8.25.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/0bcg71vhqhbq0vif5n8r93va64l8i3bl-gcc-5.4.0.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/sncl9d5v8sc5djgmpr8h7mhydz0nkyrh-gnugrep-2.25.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/53yh4wyxak4sg2j332zz4fghsg75nsn3-linux-headers-4.4.10.drv’: 1 dependencies couldn't be built
building path(s) ‘/nix/store/0djw49xzsqz9a78wl1girvx0vfmcymmm-zlib-1.2.8-static’, ‘/nix/store/2mrgjcvnmy1ha159dj1kpkq3dsxnd3pn-zlib-1.2.8’, ‘/nix/store/gxiaswl0bqnkwi8zjw56bc6k7fdh27p9-zlib-1.2.8-d
ev’
cannot build derivation ‘/nix/store/89ck7bahzlpvh8gv5vw79h7lc0d5ym08-bash-4.3-p46.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/sxfhzzdzxzx5f5w6w1ap41ibxl7vd90i-binutils-2.27.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/v4992m7w9vl2xhbg1ppbvxj9yhrniyy0-gcc-wrapper-5.4.0.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/9bpv45fh9i81zdgv65rv9yanhwdh5a2s-glibc-2.24.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/y305ppd5vrg2ij15cawy15fqca952czi-kmod-blacklist-3ubuntu1.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/dq686ia2vdziripk7vv299c125j3s4z5-lvm2-2.02.140.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/h5x35kjs05wpk69aaivn8dgsiwwys894-nixos-system-nixos-16.09beta-33749.gfedcba.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/jg600k05h91n26b7k6paz16rwrb5z5an-perl-5.22.2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/68qb6k9ns9dj5q99vczyrzkipv9zb61r-sharutils-4.11.1.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/ljrm56h1sb6m9m7rqv84yv60x10byjl0-spl-kernel-0.6.5.8-4.4.23.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/0sk8c3s57bdj4y9n9a5f7zng7fbhx1zj-stdenv.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/lwh6b1dk729maxi00xjclmha1gcwp2qa-stdenv.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/r37wlc8f4cibbpv9q6hl96vby4pdvq84-systemd-231.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/6fkshyj9dxxs9jq3gn12p8jhiwzk7b9a-zfs-kernel-0.6.5.8-4.4.23.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/94xqk0qh6b6kxbhd5hizw4rr5jpdfbn4-nixos-minimal-16.09beta-33749.gfedcba-x86_64-linux.iso.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/94xqk0qh6b6kxbhd5hizw4rr5jpdfbn4-nixos-minimal-16.09beta-33749.gfedcba-x86_64-linux.iso.drv’ failed
```
6. For your information:
```sh-session
$ ls /nix/store/...perlbug
-r-xr-xr-x 2 root root 45K Jan  1  1970 /nix/store/rcrh53q4nrdchqiwijrkxzh3v4jq5fk4-perl-5.22.2/bin/perlbug
$ git rev-parse HEAD
51d4de90c18e7541cd8b2961ef09a7094a17c938 # latest git commit that I cloned
```
This error happened before: [nixops86](https://github.com/NixOS/nixops/issues/86), [hydra129](https://github.com/NixOS/hydra/issues/129). Originally I posted it in [nixpkgs5829](https://github.com/NixOS/nixpkgs/issues/5829#issuecomment-281087262).

It would be better if we could find a faster way to reproduce this bug, because building perl takes time.  
Maybe I should try some small package -- can you offer some?

**Comment by jcumming** [comment]
I can reproduce this fairly reliably during nixos-install. I suspect the unionfs.

**Comment by sorki** [comment]
Can confirm it's unionfs issue - hitting this on netbooted hydra slave which is trying to build perl.

**Comment by FRidh** [comment]
Just got the same error on Hydra. Not sure whether it is actually related.
https://hydra.nixos.org/build/95671911

**Comment by FRidh** [comment]
Yes, this is the same issue. See https://github.com/NixOS/nixpkgs/issues/64126 and https://github.com/NixOS/nixpkgs/pull/64236

**Comment by grahamc** [comment]
FWIW these two Ni xmodules together is how I've netbooted the current aarch64 and x86 spot instance netbooted builders:


```
{ config, pkgs, lib, ... }:
let
  post-device-cmds = pkgs.writeScript "post-device-commands"
    ''
      #!/bin/sh

      set -eux
      set -o pipefail

      ${pkgs.utillinux}/bin/lsblk -d -e 1,7,11 -o NAME -n \
        | ${pkgs.busybox}/bin/sed -e "s#^#/dev/#" \
        | ${pkgs.busybox}/bin/xargs ${pkgs.zfs}/bin/zpool \
            create -f \
                -O mountpoint=none \
                -O atime=off \
                -O compression=lz4 \
                -O xattr=sa \
                -O acltype=posixacl \
                -o ashift=12 \
                rpool

      ${pkgs.zfs}/bin/zfs create -o mountpoint=legacy rpool/root

    '';
in {
  boot.supportedFilesystems = [ "zfs" ];
  boot.initrd.postDeviceCommands = "${post-device-cmds}";

  nixpkgs.config.allowUnfree = true;
  hardware.enableAllFirmware = true;
  services.openssh.enable = true;

  networking.hostId = "00000000";
}
```

^ the lsblk part above adds all disks in to the ZFS pool. Warning: this erases the disks on each boot.

Here is my modified netboot.nix:

```
{ config, lib, pkgs, ... }:

with lib;

{
  options = {

    netboot.storeContents = mkOption {
      example = literalExample "[ pkgs.stdenv ]";
      description = ''
        This option lists additional derivations to be included in the
        Nix store in the generated netboot image.
      '';
    };

  };

  config = rec {
    # Don't build the GRUB menu builder script, since we don't need it
    # here and it causes a cyclic dependency.
    boot.loader.grub.enable = false;

    # !!! Hack - attributes expected by other modules.
    environment.systemPackages = [ pkgs.grub2_efi ]
      ++ (if pkgs.stdenv.hostPlatform.system == "aarch64-linux"
          then []
          else [ pkgs.grub2 pkgs.syslinux ]);

    fileSystems."/" =
      { fsType = "zfs";
        device = "rpool/root";
      };

    fileSystems."/squash-nix-store" =
      { fsType = "squashfs";
        device = "../nix-store.squashfs";
        options = [ "loop" ];
        neededForBoot = true;
      };

    boot.initrd.postMountCommands = ''
      mkdir -p $targetRoot/nix
      cp -r $targetRoot/squash-nix-store $targetRoot/nix/store
    '';

    boot.initrd.availableKernelModules = [ "squashfs" ];

    boot.initrd.kernelModules = [ "loop" ];

    # Closures to be copied to the Nix store, namely the init
    # script and the top-level system configuration directory.
    netboot.storeContents =
      [ config.system.build.toplevel ];

    # Create the squashfs image that contains the Nix store.
    system.build.squashfsStore = pkgs.callPackage <nixpkgs/nixos/lib/make-squashfs.nix> {
      storeContents = config.netboot.storeContents;
    };


    # Create the initrd
    system.build.netbootRamdisk = pkgs.makeInitrd {
      inherit (config.boot.initrd) compressor;
      prepend = [ "${config.system.build.initialRamdisk}/initrd" ];

      contents =
        [ { object = config.system.build.squashfsStore;
            symlink = "/nix-store.squashfs";
          }
        ];
    };

    system.build.netbootIpxeScript = pkgs.writeTextDir "netboot.ipxe" ''
      #!ipxe
      kernel ${pkgs.stdenv.hostPlatform.platform.kernelTarget} init=${config.system.build.toplevel}/init ${toString config.boot.kernelParams}
      initrd initrd
      boot
    '';

    boot.loader.timeout = 10;

    boot.postBootCommands =
      ''
        # After booting, register the contents of the Nix store
        # in the Nix database in the tmpfs.
        ${config.nix.package}/bin/nix-store --load-db < /nix/store/nix-path-registration

        # nixos-rebuild also requires a "system" profile and an
        # /etc/NIXOS tag.
        touch /etc/NIXOS
        ${config.nix.package}/bin/nix-env -p /nix/var/nix/profiles/system --set /run/current-system
      '';

  };

}
```

Note: the above netboot.nix replaces the netboot.nix from the NixOS/nixpkgs repo.

Instead of running a ro/rw layer on the /nix/store, it copies /nix/store in to a FS created at `/` on the actual system disks.

**Comment by sorki** [comment]
I've switched my netbooted builders (they only have RAM) to overlayfs and it helped - See nixos/nixpkgs#35188

**Comment by sorki** [comment]
This can be probably closed since NixOS/nixpkgs#35188 is merged.
